javascript/src/jobProcessor.ts

```diff
@@ -107,95 +100,145 @@ export const jobProcessor = async (opts?: ConnectionOptions) => {
   const stopFns: StopFn[] = []
   const emitter = new EventEmitter<Events>()
   const emit = (event: Events, data: object) => {
+    debug(`${event} %O`, data)
     emitter.emit(event, { type: event, ...data })
   }
 
   const start = (def: JobDef) => {
+    emit('start', def)
     const abortController = new AbortController()
-    let timer: NodeJS.Timer
     let deferred: Deferred<void>
+    // Flag that indicates the stop function was called
+    let stopping = false
+    // Pull messages repeatedly
+    let puller: ReturnType<typeof repeater>
+    // How often to pull down messages from the consumer
+    const pullInterval = def.pullInterval ?? ms('30s')
+    // How long to wait for the batch of messages
+    const expires = pullInterval - 500
+    // Retry a failed message after a second by default
+    const backoff = def.backoff ?? ms('1s')
+    // Pull down 1 message by default
+    const batch = def.batch ?? 1
+    // Consumer config
+    const consumerConfig = consumerDefaults(def)
+
+    /**
+     * Automatically extend the ack timeout by periodically telling NATS
+     * we're working.
+     */
+    const extendAckTimeout = (msg: JsMsg) => {
+      if (def.autoExtendAckTimeout) {
+        const ackWait = consumerConfig.ack_wait
+        const intervalMs = nanosToMs(ackWait) * extendAckTimeoutThresholdFactor
+        return setInterval(() => {
+          emit('working', { ...getMetadata(msg), intervalMs })
+          msg.working()
+        }, intervalMs)
+      }
+    }
+
+    const handleTimeout = (msg: JsMsg) => {
+      const timeoutMs = def.timeoutMs
+      if (timeoutMs) {
+        return setTimeout(() => {
+          emit('timeout', { ...getMetadata(msg), timeoutMs })
+          // Abort
+          abortController.abort('timeout')
+        }, timeoutMs)
+      }
+    }
+
+    const getMetadata = (msg: JsMsg) => ({ msgInfo: msg.info, consumerConfig })
+    const areAttemptsExhausted = (msg: JsMsg) =>
+      msg.info.redeliveryCount === consumerConfig.max_deliver
+    const getExceededMs = (durationMs: number) =>
+      def.expectedMs ? durationMs - def.expectedMs : 0
```

Comment:
Does it make sense to return 0 here when the expectedMs is not exceeded (it's a negative number)

Reply:
They idea was to return `0` if you don't pass `expectedMs` since we don't know what the expected length of time is. That way it doesn't trigger alerts.

Reply:
yes, I get that part, I just thought that returning a negative number is somewhat confusing (the way the field is named currently) since we have not exceeded, we are under actually. I guess it could be useful info so that you can potentially adjust your expected time in the definition if you are consistently getting negative numbers here.

Reply:
I see. Yeah, I thought it might be useful to see the better-than-expected value represented by a negative number.